cirq-core/cirq/sim/density_matrix_simulator.py

```diff
@@ -255,49 +263,45 @@ def simulate_expectation_values_sweep(
         return swept_evs
 
 
-class DensityMatrixStepResult(simulator.StepResult['DensityMatrixSimulatorState']):
+class DensityMatrixStepResult(
+    simulator_base.StepResultBase[
+        'DensityMatrixSimulatorState', act_on_density_matrix_args.ActOnDensityMatrixArgs
+    ]
+):
     """A single step in the simulation of the DensityMatrixSimulator.
 
     Attributes:
-        qubit_map: A map from the Qubits in the Circuit to the the index
-            of this qubit for a canonical ordering. This canonical ordering
-            is used to define the state vector (see the state_vector()
-            method).
         measurements: A dictionary from measurement gate key to measurement
             results, ordered by the qubits that the measurement operates on.
     """
 
     def __init__(
         self,
-        density_matrix: np.ndarray,
-        measurements: Dict[str, np.ndarray],
-        qubit_map: Dict[ops.Qid, int],
+        sim_state: 'cirq.OperationTarget[cirq.ActOnDensityMatrixArgs]',
         dtype: 'DTypeLike' = np.complex64,
+        split_untangled_states: bool = False,
     ):
         """DensityMatrixStepResult.
 
         Args:
-            density_matrix: The density matrix at this step. Can be mutated.
-            measurements: The measurements for this step of the simulation.
-            qubit_map: A map from qid to index used to define the
-                ordering of the basis in density_matrix.
-            dtype: The numpy dtype for the density matrix.
+            sim_state: The qubit:ActOnArgs lookup for this step.
+            dtype: The `numpy.dtype` used by the simulation. One of
+                `numpy.complex64` or `numpy.complex128`.
         """
-        super().__init__(measurements)
-        self._density_matrix = density_matrix
-        self._qubit_map = qubit_map
+        super().__init__(sim_state)
         self._dtype = dtype
-        self._qid_shape = simulator._qubit_map_to_shape(qubit_map)
-
-    def _qid_shape_(self):
-        return self._qid_shape
+        self._density_matrix: Optional[np.ndarray] = None
+        self._split_untangled_states = split_untangled_states
 
     def _simulator_state(self) -> 'DensityMatrixSimulatorState':
-        return DensityMatrixSimulatorState(self._density_matrix, self._qubit_map)
+        return DensityMatrixSimulatorState(self.density_matrix(copy=False), self._qubit_mapping)
 
     def set_density_matrix(self, density_matrix_repr: Union[int, np.ndarray]):
```

Comment:
Make sure this is fixed per https://github.com/quantumlib/Cirq/pull/4100#issuecomment-864257447